### ISSUE TITLE
Private Method Support in Interface Mixins

### ISSUE THREAD

**Issue by LambdAurora** [FabricMC/Mixin#54]
From SpongePowered/Mixin#497:

---

Mixin and interfaces is a long story of friendship.

Let's take it from the start and talk about the `Oxidizable` interface in Minecraft 1.17, all names here are from the yarn mappings.

All oxidation level increases are done with a supplier of a bimap of blocks, and the level decreases one is simply the inversion of the first.

It looks like that
```java
Supplier<BiMap<Block, Block>> OXIDATION_LEVEL_INCREASES = Suppliers.memoize(() -> { return /* an immutable bi map builder here */; });
```

For any mods which wants to add blocks in this map will find a wall, a big one.

Let's take a look at Java 8 first and the issues:

 - a setter accessor is impossible due to the fact that it would try to remove the `final` and the JVM won't like that at all.
 - an injector to the lambda is impossible due to the fact that: 
   - mixin will refuse anything that isn't an interface to mixin an interface
   - mixin will refuse to use a `public static` injector (required by interfaces in Java 8)
 - any other injection-like will fail due to the same issues.
 
 With the recent upgrade to Java 16 we now have access to private methods in interfaces, and I guessed I could finally ditch the ugly raw-asm transformation, but yet it's still impossible due to the fact that mixin yells ` tesla_coil.mixins.json:OxidizableMixin: Interface mixin contains a non-public method! Found onBuildLevelIncreasesMap(Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;)V in tesla_coil.mixins.json:OxidizableMixin`
 
 So, I cannot use a class mixin, I cannot use a public static injector, nor a private static injector.
 
 (note: even though Java 16 is not yet officially supported, the issue still exist if using Java 11 as the private methods in interfaces are in Java since Java 9)
 
 To conclude, mixin is way too aggressive in its checks globally which impedes its capacities.
 
---

It's been a while now and upstream still has nothing for that.

This is becoming quite an issue with 1.17, with issues rising and mod compat being hurt.

Please consider fixing it in this fork for modders' sake.

**Comment by Devan-Kerman** [comment]
https://github.com/FabricMC/Mixin/pull/51 would fix this right?

**Comment by LambdAurora** [comment]
> #51 would fix this right?

Seems like it. With 1.17 this kinda becomes mandatory for Oxidization (literally no work around exist, except Mixin Plugin + Raw ASM), and tooltip components.

**Comment by Devan-Kerman** [comment]
also, does the JVM complain when a static variable in an interface is non-final? When I tried using an accessor mixin to access oxidizable, I got an unrelated error

**Comment by Chocohead** [comment]
It's not a fan:
```
java.lang.ClassFormatError: JVMCFRE009 interface field must be public static and final; class=com/chocohead/test/Interface, offset=1765
		at java.base/java.lang.ClassLoader.defineClassImpl(Native Method)
		at java.base/java.lang.ClassLoader.defineClassInternal(ClassLoader.java:482)
		at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:451)
		at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:151)
		at net.fabricmc.loader.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:156)
		at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:1055)
		at java.base/java.lang.ClassLoader.defineClassImpl(Native Method)
		at java.base/java.lang.ClassLoader.defineClassInternal(ClassLoader.java:482)
		at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:451)
		at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:151)
		at net.fabricmc.loader.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:156)
		at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:1055)
		at io.github.minecraftcursedlegacy.example.ExampleMod.testInterface(ExampleMod.java:46)
		at io.github.minecraftcursedlegacy.example.ExampleMod.onInitialize(ExampleMod.java:26)
		at net.fabricmc.loader.entrypoint.minecraft.hooks.EntrypointClient$$Lambda$166.000000009D723DB0.accept(Unknown Source)
		at net.fabricmc.loader.entrypoint.minecraft.hooks.EntrypointUtils.invoke0(EntrypointUtils.java:47)
		... 16 more
```
Of course this only happens when you annotate your setter with `@Mutable`, you can ignore it but then you're trusting the VM is happy for you to set a static final field

**Comment by Devan-Kerman** [comment]
TIL, guess I just never tried it before

**Comment by LambdAurora** [comment]
Fixed via #51.
